File: Demo/SLDSSampleApp/ColorListViewController.m
/*
 Copyright (c) 2015, salesforce.com, inc. All rights reserved.
 Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
 Neither the name of salesforce.com, inc. nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "ColorListViewController.h"
#import "ColorViewController.h"
#import <SalesforceDesignSystem/SalesforceDesignSystemExtended.h>
#import "SWRevealViewController.h"


@interface ColorListViewController ()

@end

@implementation ColorListViewController {
    NSArray *textColors;
    NSArray *backgroundColors;
    NSArray *borderColors;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self navigationItem].title = @"Landmark Icons";
    

    NSMutableArray *tColors = [[NSMutableArray alloc] init];
    for(int i=SLDSColorTextActionLabel; i<=SLDSColorTextWeak; i++) {
        [tColors addObject:[NSNumber numberWithInt:i]];
    }
    textColors = [tColors copy];

    NSMutableArray *bColors = [[NSMutableArray alloc] init];
    for(int i=SLDSColorBackground; i<=SLDSColorBackgroundTempModalTint; i++) {
        [bColors addObject:[NSNumber numberWithInt:i]];
    }
    backgroundColors = [bColors copy];

    
    NSMutableArray *brdrColors = [[NSMutableArray alloc] init];
    for(int i=SLDSColorBorder; i<=SLDSColorBorderWarning; i++) {
        [brdrColors addObject:[NSNumber numberWithInt:i]];
    }
    borderColors = [brdrColors copy];

    
    SWRevealViewController *revealController = self.revealViewController;
    
    [self.view addGestureRecognizer:revealController.panGestureRecognizer];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSArray *) sourceForSection:(NSInteger)section {
    if(section == 0){
        return textColors;
    }
    else if(section == 1){
        return backgroundColors;
    }
    else if(section == 2){
        return borderColors;
    }
    return @[];
}

- (NSString *) getLabel:(NSIndexPath *)indexPath{
    switch (indexPath.section) {
        case 0:
            return [NSString sldsColorTextName:[[[self sourceForSection:indexPath.section] objectAtIndex:indexPath.item] intValue]];
        case 1:
            return [NSString sldsColorBackgroundName:[[[self sourceForSection:indexPath.section] objectAtIndex:indexPath.item] intValue]];
        case 2:
            return [NSString sldsColorBorderName:[[[self sourceForSection:indexPath.section] objectAtIndex:indexPath.item] intValue]];
        default:
            return @"NOT FOUND!";
    }
}

- (UIColor *) getColor:(NSIndexPath *)indexPath{
    switch (indexPath.section) {
        case 0:
            return [UIColor sldsColorText:[[[self sourceForSection:indexPath.section] objectAtIndex:indexPath.item] intValue]];
        case 1:
            return [UIColor sldsColorBackground:[[[self sourceForSection:indexPath.section] objectAtIndex:indexPath.item] intValue]];
        case 2:
            return [UIColor sldsColorBorder:[[[self sourceForSection:indexPath.section] objectAtIndex:indexPath.item] intValue]];
        default:
            return [UIColor clearColor];
    }
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 3;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [[self sourceForSection:section] count];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    if(section == 0){
        return @"TEXT COLORS";
    }
    if(section == 1){
        return @"BACKGROUND COLORS";
    }
    if(section == 2){
        return @"BORDER COLORS";
    }
    return @"";
}

- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section
{
    view.tintColor = [UIColor whiteColor];
    
    UITableViewHeaderFooterView *header = (UITableViewHeaderFooterView *)view;
    [header.textLabel setFont:[UIFont sldsFontLightWithSize:SLDSFontSizeLarge]];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *simpleTableIdentifier = @"LMColorCell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:simpleTableIdentifier];
    
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:simpleTableIdentifier];
    }
    
    [cell.textLabel setFont:[UIFont sldsFontRegularWithSize:SLDSFontSizeSmall]];
    
    cell.textLabel.text = [self getLabel:indexPath];
    
    cell.backgroundColor = [self getColor:indexPath];
    
    return cell;
    
}

- (CGFloat)tableView:(UITableView *)t heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 50;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
        SWRevealViewController *revealController = self.revealViewController;

        ColorViewController *dc = [[ColorViewController alloc] init];
        
        dc.colorID = [[[self sourceForSection:indexPath.section] objectAtIndex:indexPath.item] integerValue];
        dc.colorCategoryID = [self categoryId:indexPath];
    
    [revealController setFrontViewController:dc animated:YES];
    [revealController setFrontViewPosition:FrontViewPositionLeftSideMost animated:YES];
 }

-(LM_COLOR_CATEGORY)categoryId:(NSIndexPath *) indexPath{
    switch(indexPath.section){
        case 0: return LM_COLOR_CATEGORY_TEXT;
        case 1: return LM_COLOR_CATEGORY_BACKGROUND;
        case 2: return LM_COLOR_CATEGORY_BORDER;
    }
    return LM_COLOR_CATEGORY_BACKGROUND;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([[segue identifier] isEqualToString:@"ShowDetailView"]) {
        ColorViewController *dc = [segue destinationViewController];
        
        dc.colorID = [[(NSArray *)sender objectAtIndex:1] intValue];
        dc.colorCategoryID = [[(NSArray *)sender objectAtIndex:0] intValue];
    }
}


@end

File: Demo-Swift/slds-sample-app/ColorListViewController.swift
/*
 Copyright (c) 2016, salesforce.com, inc. All rights reserved.
 Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
 Neither the name of salesforce.com, inc. nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import UIKit

//––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––

enum ColorListType : String {
    case background = "Background"
    case border = "Border"
    case text = "Text"
}

struct ColorObject {
    var color : UIColor!
    var alias : String!
    var method : String!
}

//––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––

class ColorListViewController: UITableViewController {
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    var colors : [ColorObject] = [ColorObject](){
        didSet {
            colors.sort { (c1, c2) -> Bool in
                var h1:CGFloat = 0.0
                var s1:CGFloat = 0.0
                var b1:CGFloat = 0.0
                var a1:CGFloat = 0.0
                
                var h2:CGFloat = 0.0
                var s2:CGFloat = 0.0
                var b2:CGFloat = 0.0
                var a2:CGFloat = 0.0
                
                c1.color.getHue(&h1, saturation: &s1, brightness: &b1, alpha: &a1)
                c2.color.getHue(&h2, saturation: &s2, brightness: &b2, alpha: &a2)
                
                if h1 == h2 {
                    if s1 == s2 {
                        if b1 == b2 {
                            return a2 < a1
                        } else {
                            return b1 < b2
                        }
                    } else {
                        return s1 < s2
                    }
                } else {
                    return h1 < h2
                }
            }
        }
    }
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    override var title: String? {
        didSet {
            super.title = self.title
            if let newTitle = self.title {
                self.addContent(newTitle)
            }
        }
    }
    
    
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    override func viewDidLoad() {
        self.tableView.register(ColorCell.self, forCellReuseIdentifier: "Cell")
        self.tableView.register(AccountCell.self, forCellReuseIdentifier: "Cell2")
    }
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> ColorCell {
        
        let alias = colors[indexPath.item].alias
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath) as! ColorCell
        
        cell.backgroundColor = UIColor.clear
        cell.textLabel?.text = alias?.replacingOccurrences(of: "SLDSColor", with: "")
        cell.textLabel?.font = UIFont.sldsFont(.regular, with: .small)
        cell.dataProvider = colors[indexPath.item].color
        
        return cell
    }
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 100
    }
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return colors.count
    }
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let controller = ColorViewController()
        controller.dataProvider = self.colors[indexPath.row]
        self.navigationController?.show(controller, sender: self)
    }
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    func addContent(_ colorType: String) {
        
        var colorList = Array<ColorObject>()
        var indexPath : IndexPath
        
        switch colorType {
        case ColorListType.background.rawValue : indexPath = IndexPath(row: 0, section: 0)
        case ColorListType.border.rawValue : indexPath = IndexPath(row: 0, section: 1)
        case ColorListType.text.rawValue : indexPath = IndexPath(row: 0, section: 2)
        default : return
        }
        
        while let color = self.getColorAt(indexPath) {
            colorList.append(color)
            indexPath.row = colorList.count
        }
        
        self.colors = colorList
    }
    
    //––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––––
    
    func getColorAt(_ indexPath: IndexPath) -> ColorObject? {
        
        var retVal : ColorObject?
        
        switch indexPath.section {
            case 0 :
                if indexPath.row > 0 && (SLDSColorBackgroundType.init(rawValue: indexPath.row)?.hashValue == 0) {
                    break
                }
                
                if let value = SLDSColorBackgroundType.init(rawValue: indexPath.row) {
                    retVal = ColorObject(color: UIColor.sldsColorBackground(value),
                                         alias: NSString.sldsColorBackgroundName(value) as String,
                                         method: "sldsColorBackground")
                }
            
            case 1 :
                if indexPath.row > 0 && (SLDSColorBorderType.init(rawValue: indexPath.row+1)?.hashValue == 0) {
                    break
                }
                if let value = SLDSColorBorderType.init(rawValue:  indexPath.row) {
                    retVal = ColorObject(color: UIColor.sldsColorBorder(value),
                                         alias: NSString.sldsColorBorderName(value) as String,
                                         method: "sldsColorBorder")
                }
            
            case 2 :
                if indexPath.row > 0 && (SLDSColorTextType.init(rawValue: indexPath.row)?.hashValue == 0) {
                    break
                }
                
                if let value = SLDSColorTextType.init(rawValue:  indexPath.row) {
                    retVal = ColorObject(color: UIColor.sldsColorText(value),
                                         alias: NSString.sldsColorTextName(value) as String,
                                         method: "sldsColorText")
                }
            
            default : break
        }
        return retVal
    }
}
